pallets/axelar-router/src/lib.rs

```diff
@@ -0,0 +1,419 @@
+	/// contract code is used.
+	pub target_contract_hash: H256,
+
+	/// The values used when executing the EVM call.
+	pub fee_values: FeeValues,
+}
+
+/// The FeeValues holds all information related to the transaction costs.
+#[derive(Debug, Encode, Decode, Clone, PartialEq, Eq, TypeInfo, MaxEncodedLen)]
+pub struct FeeValues {
+	/// The value used when executing the EVM call.
+	pub value: U256,
+
+	/// The gas price used when executing the EVM call.
+	pub gas_price: U256,
+
+	/// The gas limit used when executing the EVM call.
+	pub gas_limit: U256,
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// The origin that is allowed to set the gateway address we accept
+		/// messages from
+		type AdminOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+
+		/// The target of the messages comming from other chains
+		type Receiver: MessageReceiver<Middleware = Self::Middleware, Origin = DomainAddress>;
+
+		/// Middleware used by the gateway
+		type Middleware: From<AxelarId>;
+
+		/// The target of the messages comming from this chain
+		type Transactor: EthereumTransactor;
+
+		/// Checker to ensure an evm account code is registered
+		type EvmAccountCodeChecker: PreConditions<(H160, H256), Result = bool>;
+	}
+
+	#[pallet::storage]
+	pub type Configuration<T: Config> = StorageMap<_, Twox64Concat, ChainName, AxelarConfig>;
+
+	#[pallet::storage]
+	pub type ChainNameById<T: Config> = StorageMap<_, Twox64Concat, AxelarId, ChainName>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		ConfigSet {
+			name: ChainName,
+			config: Box<AxelarConfig>,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Emit when the router configuration is not found.
+		RouterNotFound,
+
+		/// Emit when the evm account code is not registered
+		ContractCodeMismatch,
+
+		/// Emit when the source chain is too big
+		SourceChainTooLong,
+
+		/// Emit when the source address can not be recognized
+		InvalidSourceAddress,
+
+		/// Emit when a message is received from a non LP caller
+		ContractCallerMismatch,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(Weight::from_parts(50_000_000, 512).saturating_add(RocksDbWeight::get().writes(2)))]
+		#[pallet::call_index(0)]
+		pub fn set_config(
+			origin: OriginFor<T>,
+			chain_name: ChainName,
+			config: Box<AxelarConfig>,
+		) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+
+			match &config.domain {
+				DomainConfig::Evm(evm_config) => {
+					ensure!(
+						T::EvmAccountCodeChecker::check((
+							evm_config.target_contract_address,
+							evm_config.target_contract_hash,
+						)),
+						Error::<T>::ContractCodeMismatch
+					);
+
+					ChainNameById::<T>::insert(
```

Comment:
Do we really need a separate storage for the chain name? Why not make that part of the `AxelarConfig`?

Reply:
We need it because for sending we index by `AxelarId` but for receiving we index by `ChainName` (basically, the latest is an Axelar requirement in their API). So we have two identifications that redirect to the same config.

Reply:
Ah, I looked at that but didn't notice it, maybe I need to start drinking coffee ^^. My bad, sorry for the noise.

Reply:
No worries! good question anyways 🫶